### DISCUSSION TITLE
Model

### DISCUSSION THREAD

**Discussion by accuser** [swe-delivery/swe-tube#7] [Ideas]
```mermaid
classDiagram
  class User {
    +id: int
    +email: string
    +password: string
    +username: string
    +is_teacher: boolean
    +created_at: datetime
    +updated_at: datetime
  }

  class Resource {
    +id: int
    +title: string
    +description: string
    +url: string
    +type: string
    +created_by: User
    +created_at: datetime
    +updated_at: datetime
  }

  class Category {
    +id: int
    +name: string
    +created_by: User
    +created_at: datetime
    +updated_at: datetime
  }

  class Tag {
    +id: int
    +name: string
    +created_by: User
    +created_at: datetime
    +updated_at: datetime
  }

  class ResourceCategory {
    +id: int
    +resource_id: int
    +category_id: int
  }

  class ResourceTag {
    +id: int
    +resource_id: int
    +tag_id: int
  }

  class Recommendation {
    +id: int
    +resource_id: int
    +recommended_by: User
    +created_at: datetime
    +updated_at: datetime
  }

  User "1" -- "*" Resource : creates
  User "1" -- "*" Category : creates
  User "1" -- "*" Tag : creates
  Resource "1" -- "*" ResourceCategory : categorized as
  Resource "1" -- "*" ResourceTag : tagged as
  User "1" -- "*" Recommendation : recommends
  Resource "1" -- "*" Recommendation : recommended for
```

With this data model, we can easily query for resources based on their `type`, `category`, or `tag`, and we can also see which resources have been recommended by other users. The `ResourceCategory` and `ResourceTag` models allow us to link resources to one or more categories or tags, respectively. The `Recommendation` model allows users to give feedback on resources and helps other users find high-quality content. The `User` model distinguishes between teachers and students (although I wonder if single table inheritance approach my be better), so that we can control who has the ability to add new resources and moderate student submissions.

**Comment by nimonian** [comment]
Sensational

Is a "recommendation" the same as a "like"? Or would a user recommend something _to_ another user?

**Reply by nimonian** [reply]
Also I totally added a `spec.md` and then deleted it -- ignore it, it was before I even saw that discussions had been ongoing

**Reply by accuser** [reply]
This were based on the user stories as a starting point. A recommendation could be a like.

**Comment by nimonian** [comment]
A meta-comment:

In terms of data modelling for the apprentices, they experience an unpleasant dissonance when they pass from week 5 (OOP) to week 6 (databases). They want to use a diagram like what you've drawn above. But when they're making data models and tables, they aren't using the `class` keyword, and so they get stuck with it (how can I use a Class Diagram when I'm not using actual classes?).

What you've got here is essentially and entity relationship diagram (i.e. a class diagram without the methods). I think the curriculum should underline that this is a use case for class diagrams: we can use them to model out our system even if the entities don't correspond to actual "classes" in our code.

Does this make sense? It was a weird blocker for apprentices

**Reply by accuser** [reply]
I think there is something about encouraging them to step back and see the bigger picture. A data model doesn't necessarily map to a database, it is simply how we want to model our domain of concern. We may well put some of this in a database, but that's really an implementation detail - it should only matter to that part of the system that deals with persistence.

 Same way we need to help apprentices stop thinking that REST maps to CRUD.

**Reply by accuser** [reply]
Separating concerns is what we should be encouraging; thinking about software as cooperating systems. As a developer I know I have a database, but I don't want my REST API to know that. It shouldn't care. It should only care that when a `GET` request with an associated `id` route parameter invokes the right handler, the handler can delegate the getting of the resource and return the right response.

**Reply by Jaardy** [reply]
> I think there is something about encouraging them to step back and see the bigger picture. A data model doesn't necessarily map to a database, it is simply how we want to model our domain of concern. We may well put some of this in a database, but that's really an implementation detail - it should only matter to that part of the system that deals with persistence.
> 
> Same way we need to help apprentices stop thinking that REST maps to CRUD.

From a teaching perspective, would you highlight this as the "interface" point for our app?  A domain of concern suggests boundaries, the clearer the better!

**Reply by accuser** [reply]
The domain of concern is about the system's world view - what it knows about what it cares about. In this case, it's the sort of things described at the top of the thread. The system might provide mechanisms to link to YouTube video resources, but it doesn't care that they are YouTube video resources - all it cares is that it has a resource with a URL. It is a constraining world view.

The interface is different (although I appreciate the interface will be informed by the domain of concern) but it does provide a boundary. Consider the backend as service implementing a RESTful API. You can draw a box around the backend, and look at the edges: what messages are coming in (HTTP Request) and what messages are going out (HTTP Response). Everything inside the box is a mystery, but we know how to talk to it. The boundary is well defined.

**Reply by Jaardy** [reply]
Yes, exactly this.  With regards to the interface, the analogy I used that seemed most successful was that of the wiring in your house and the outlet and the plugs being your connecting "interfaces"  There is no need for us to go beyond that boundary (ie open up and hardwire into the circuit.  From the system's side, it doesn't really know or care what you plug in.   It does require the correct interface (HTTP, FTP, TCP etc) and there is something similar in the idea of the api contract.  I'm thinking of circuit breakers being there to protect the server.
Maybe extending this analogy a bit further, the electricity in the house doesn't know where its sourcing its power from: coal, nuclear, wind.
A recent example where apprentices struggled with these boundaries was with the idea of user authentication with tokens.  The idea they expressed was that "now that the user is logged in" after visiting the login end-point and successfully returning a signed token, that future  request/response cycles would somehow be different.  As far as area of concerns go, I had to highlight that the api service has returned the token and essential has no memory of the transaction.  It is the concern of the client to manage that token.  There is also no way for the server to "log the user out" short of having the token expire.  The server is basically a "10 second Tom."

Apologies for the wall of text.    I find this helpful from the perspective of one who is delivering to apprentices, and building in these analogies for future instructors as well.  I just don't want to lay any bear traps with a particular mental model.

**Reply by nimonian** [reply]
Really like the house wiring analogy - really helps to understand these things in terms of what we already know.

I definitely think we can do a better job of explaining not only what out REST API _is_, but also where it sits in relation to everything else in the system.
